File: python/802. Find Eventual Safe States.py
class Solution(object):
    def eventualSafeNodes(self, graph):
        """
        :type graph: List[List[int]]
        :rtype: List[int]
        """
        n = len(graph)
        vis = [0]*n 
        def dfs(cur):
            if vis[cur]==-1:
                return False
            if vis[cur]==1:
                return True
            vis[cur] = -1
            for i in graph[cur]:
                if not dfs(i):
                    return False
            vis[cur] = 1
            return True
        
        for i in range(n):
            if not vis[i]:
                dfs(i)
        return [i for i in range(n) if vis[i]==1]

# 寻找不在环中的结点

File: go/src/leetcode/802. Find Eventual Safe States/802. Find Eventual Safe States.go
package main

import "fmt"

/**
https://leetcode-cn.com/problems/find-eventual-safe-states/
思路：dfs, 剪枝

 */
func eventualSafeNodes(graph [][]int) []int {
	n := len(graph)
	circle := make([]int, n) // (-1, 0, 1) 分别表示在环中、未访问、不在环中

	var dfs func(idx int) bool
	dfs = func(idx int) bool {
		if circle[idx] == -1 {
			return false
		}
		if circle[idx] == 1 {
			return true
		}
		circle[idx] = -1
		for _, ni := range graph[idx] {
			if !dfs(ni) {
				return false
			}
		}
		circle[idx] = 1
		return true
	}

	for i := 0; i < n; i++ {
		if circle[i] == 0 {
			dfs(i)
		}
	}

	var ans = []int{}

	for i := 0; i < n; i++ {
		if circle[i] == 1 {
			ans = append(ans, i)
		}
	}

	fmt.Println(ans)
	return ans
}

func main() {
	eventualSafeNodes([][]int{{1,2},{2,3},{5},{0},{5},{},{}})
	eventualSafeNodes([][]int{{1,2,3,4},{1,2},{3,4},{0,4},{}})
	eventualSafeNodes([][]int{{0},{2,3,4},{3,4},{0,4},{}})
}


